tools/ci/run_tc.py

```diff
@@ -262,13 +266,60 @@ def setup_environment(args):
     if args.oom_killer:
         start_userspace_oom_killer()
 
-    if args.checkout:
-        checkout_revision(args.checkout)
 
+def setup_repository(args):
+    is_pr = os.environ.get("GITHUB_PULL_REQUEST", "false") != "false"
+
+    # Initially task_head points at the same commit as the ref we want to test
+    # However that may not be the same commit as we actually want to test if
+    # the branch changed since the decision task ran.
```

Comment:
I think understanding this is crucial to me being able to review this patch. What is the actual race condition we're trying to deal with here; why could the branch change since the decision task ran? 

In terms of what information taskcluster needs, I'm thinking we want to be able to run browsers before and after the change (in the case of pull requests), or otherwise just after the change (in the case of commits aka pushes). That seems like 2 SHAs for a PR and 1 SHA for a commit to me, but I think there's definitely some information I'm missing here :)

Reply:
So you push to a branch (or a PR), a decision task runs, and at some later time the scheduled tasks run. In the meantime there may have been more commits that affect the content of that push; most trivially someone may have landed another commit on top of the one that you were trying to test. For PRs, where we are testing an equivalent merge commit, the situation is even worse, either a push to the head or a push to base between the decision task running and the subsequent tasks running can change what the branch ref points at.

So one obvious solution here would be to always get the relevant SHAs out of the GitHub event and just pass the SHA around. But that doesn't work well; in general you can't fetch an object by SHA in git unless it happens that it's referred to by some ref. For runs based on branches that's not a huge problem because typically commits don't go away (although force pushing to trigger branches is an example of where this might not be true). But for the merge commits in PRs, the original commit can go away just because the base branch changes, which shouldn't be enough to stop us running tests.

So the previous solution here was to pass around the branch name to fetch from and the expected SHA at the head of the branch, so we could fail gracefully in the case where the branch was updated, and notably in the case it was force pushed. But that also didn't account for the fact that for PRs we want to handle changes to the PR head differently from changes to the base; we want to abort for the former and test anyway for the latter.

So with this PR we pass around the branch head, the base and the branch name and try to handle each condition explicitly.

Does that help?

Reply:
That is an amazing description, thank you! My understanding has gone up 100x! Will start reviewing now, and I may ask for some of this amazing explanation to be incorporated into comments :).